src/app/clusters/thermostat-server/thermostat-server-presets.cpp

```diff
@@ -402,6 +402,28 @@ CHIP_ERROR ThermostatAttrAccess::AppendPendingPreset(Thermostat::Delegate * dele
         return CHIP_IM_GLOBAL_STATUS(ConstraintError);
     }
 
+    // Before adding this preset to the pending presets, if the expected length of the pending presets' list
+    // exceeds the total number of presets supported, return RESOURCE_EXHAUSTED. Note that the preset has not been appended yet.
+
+    // Increment number of pending presets by 1 to account for this preset.
+    uint8_t totalExpectedCount = CountNumberOfPendingPresets(delegate) + 1;
+
+    uint8_t numberOfPresetsSupported = delegate->GetNumberOfPresets();
+
+    if (numberOfPresetsSupported == 0)
+    {
+        ChipLogError(Zcl, "AppendPendingPreset: Failed to get NumberOfPresets");
+        return CHIP_IM_GLOBAL_STATUS(InvalidInState);
+    }
+
+    if (numberOfPresetsSupported > 0 && totalExpectedCount > numberOfPresetsSupported)
```

Comment:
Comment got lost in the rebase, apparently: this is an unsigned value, and you just checked if it was zero on 413, so the first part of this is always true.

Reply:
fixed.